### ISSUE TITLE
Incompatible versions in Cirq 1.0.0

### ISSUE THREAD

**Issue by ianhellstrom** [quantumlib/Cirq#5929]
**Description of the issue**
Protobuf versions of various transient dependencies do no match for Cirq 1.0.0. For those of us who prefer to peg versions, this represents a problem.

**How to reproduce the issue**

1. Create reqs.in file with `cirq==1.0.0` as its contents
2. Run `pip-compile reqs.in`
3. Check output:

```
Could not find a version that matches protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<4,<4.0.0dev,<5.0.0dev,>=3.15.0,>=3.19.0,>=3.19.5,>=4.21.6 (from cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
Tried: 2.0.3, 2.3.0, 2.4.1, 2.5.0, 2.6.0, 2.6.1, 3.0.0, 3.0.0, 3.1.0, 3.1.0.post1, 3.1.0.post1, 3.2.0, 3.2.0, 3.3.0, 3.4.0, 3.4.0, 3.5.0.post1, 3.5.0.post1, 3.5.1, 3.5.1, 3.5.2, 3.5.2, 3.5.2.post1, 3.5.2.post1, 3.6.0, 3.6.0, 3.6.1, 3.6.1, 3.7.0, 3.7.0, 3.7.1, 3.7.1, 3.8.0, 3.8.0, 3.9.0, 3.9.0, 3.9.1, 3.9.1, 3.9.2, 3.9.2, 3.10.0, 3.10.0, 3.11.0, 3.11.0, 3.11.1, 3.11.1, 3.11.2, 3.11.2, 3.11.3, 3.11.3, 3.12.2, 3.12.2, 3.12.4, 3.12.4, 3.13.0, 3.13.0, 3.14.0, 3.14.0, 3.15.0, 3.15.0, 3.15.1, 3.15.1, 3.15.2, 3.15.2, 3.15.3, 3.15.3, 3.15.4, 3.15.4, 3.15.5, 3.15.5, 3.15.6, 3.15.6, 3.15.7, 3.15.7, 3.15.8, 3.15.8, 3.16.0, 3.16.0, 3.17.0, 3.17.0, 3.17.1, 3.17.1, 3.17.2, 3.17.2, 3.17.3, 3.17.3, 3.18.0, 3.18.0, 3.18.1, 3.18.1, 3.18.3, 3.18.3, 3.19.0, 3.19.0, 3.19.1, 3.19.1, 3.19.2, 3.19.2, 3.19.3, 3.19.3, 3.19.4, 3.19.4, 3.19.5, 3.19.5, 3.19.6, 3.19.6, 3.20.0, 3.20.0, 3.20.1, 3.20.1, 3.20.2, 3.20.2, 3.20.3, 3.20.3, 4.21.0, 4.21.0, 4.21.0, 4.21.0, 4.21.1, 4.21.1, 4.21.1, 4.21.1, 4.21.2, 4.21.2, 4.21.2, 4.21.2, 4.21.3, 4.21.3, 4.21.3, 4.21.3, 4.21.4, 4.21.4, 4.21.4, 4.21.4, 4.21.5, 4.21.5, 4.21.5, 4.21.5, 4.21.6, 4.21.6, 4.21.6, 4.21.6, 4.21.7, 4.21.7, 4.21.7, 4.21.7, 4.21.8, 4.21.8, 4.21.8, 4.21.8
Skipped pre-versions: 2.0.0b0, 3.0.0a2, 3.0.0a3, 3.0.0b1, 3.0.0b1.post1, 3.0.0b1.post2, 3.0.0b2, 3.0.0b2, 3.0.0b2.post1, 3.0.0b2.post1, 3.0.0b2.post2, 3.0.0b2.post2, 3.0.0b3, 3.0.0b4, 3.0.0b4, 3.2.0rc1, 3.2.0rc1, 3.2.0rc1.post1, 3.2.0rc1.post1, 3.2.0rc2, 3.2.0rc2, 3.7.0rc2, 3.7.0rc2, 3.7.0rc3, 3.7.0rc3, 3.8.0rc1, 3.8.0rc1, 3.9.0rc1, 3.9.0rc1, 3.10.0rc1, 3.10.0rc1, 3.11.0rc1, 3.11.0rc1, 3.11.0rc2, 3.11.0rc2, 3.13.0rc3, 3.13.0rc3, 3.14.0rc1, 3.14.0rc1, 3.14.0rc2, 3.14.0rc2, 3.14.0rc3, 3.14.0rc3, 3.15.0rc1, 3.15.0rc1, 3.15.0rc2, 3.15.0rc2, 3.16.0rc1, 3.16.0rc1, 3.16.0rc2, 3.16.0rc2, 3.17.0rc1, 3.17.0rc1, 3.17.0rc2, 3.17.0rc2, 3.18.0rc1, 3.18.0rc1, 3.18.0rc2, 3.18.0rc2, 3.19.0rc1, 3.19.0rc1, 3.19.0rc2, 3.19.0rc2, 3.20.0rc1, 3.20.0rc1, 3.20.0rc2, 3.20.0rc2, 3.20.1rc1, 3.20.1rc1, 4.0.0rc1, 4.0.0rc1, 4.0.0rc2, 4.0.0rc2, 4.21.0rc1, 4.21.0rc1, 4.21.0rc1, 4.21.0rc1, 4.21.0rc2, 4.21.0rc2, 4.21.0rc2, 4.21.0rc2
There are incompatible versions in the resolved dependencies:
  protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<4.0.0dev,>=3.19.5 (from google-api-core[grpc]==1.33.2->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
  protobuf<4,>=3.15.0 (from cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
  protobuf<5.0.0dev,>=3.19.0 (from proto-plus==1.22.1->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
  protobuf<5.0.0dev,>=3.15.0 (from googleapis-common-protos==1.56.4->google-api-core[grpc]==1.33.2->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
  protobuf>=4.21.6 (from grpcio-status==1.50.0->google-api-core[grpc]==1.33.2->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))
```

**Cirq version**
1.0.0

**Comment by dstrain115** [comment]
Discussion from Cirq Cync: Pavol to investigate and determine whether this issue is reproducible and what the current status of protobuf libraries and downstream dependencies are.

**Comment by mpharrigan** [comment]
+1 this issue. `pip-compile` is becoming an important part of the python ecosystem and does not work whenever cirq is involved

**Comment by maffoo** [comment]
Thanks for looking into this, @pavoljuhas. If I'm reading the output above correctly, this looks like a problem with google-api-core. The first line says:

> protobuf!=3.20.0,!=3.20.1,!=4.21.0,!=4.21.1,!=4.21.2,!=4.21.3,!=4.21.4,!=4.21.5,<4.0.0dev,>=3.19.5 (from google-api-core[grpc]==1.33.2->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))

That's a bit hard to parse, but the `<4.0.0dev` looks like `google-api-core[grpc]` want proto 3. However the last line

> protobuf>=4.21.6 (from grpcio-status==1.50.0->google-api-core[grpc]==1.33.2->cirq-google==1.0.0->cirq==1.0.0->-r reqs.in (line 1))

Makes it look like google-api-core is pulling in a version of `grpcio-status` that wants a newer proto version. Can we fix this by doing something like pinning the `grpcio-status` version to one that is compatible with older protobuf? That would be a less drastic solution that updating to protobuf 4 if we can make it work.

**Comment by pavoljuhas** [comment]
> Can we fix this by doing something like pinning the grpcio-status version to one that is compatible with older protobuf? That would be a less drastic solution that updating to protobuf 4 if we can make it work.

Good point, please see #5953.

**Comment by pavoljuhas** [comment]
@ianhellstrom - it turns out that cirq dependencies can be successfully resolved by using
`pip-compile --resolver=backtracking`

The backtracking resolver will become the new default for pip-compile.
Without the `--resolver` option the pip-compile shows a warning to that effect:
> WARNING: using legacy resolver is deprecated and will be removed in future versions. The default resolver will be change to 'backtracking' in 7.0.0 version. Specify --resolver=backtracking to silence this warning.

Hope this works for getting pip-compile run.  Please reopen if there are still problems.

Note: #5953 has a fix for legacy resolver, but we chose not to merge, because `--resolver=backtracking` produces the same output.
